### PR TITLE
Fix junk data in ffprobe

### PR DESCRIPTION
**Pull request by Marekkon5**
Hello, this is my attempt at fixing issue #39 .

I've made a proper file using Kid3 and then compared it to tag from this library and made the following changes:
- The padding in file should be 1024, however your code somewhere adds an extra byte, so made it 1023
- The header size (and 1 extra byte) isn't included in the good tag, so removed them, however the 1024b padding is included.

I've also added a `quiet.mp3` file, which after tagging works in ffprobe and didn't without my changes.

Thank you.

### CODE REVIEW

File: src/tag.rs [hunk 1]
```diff
@@ -1425,4 +1428,12 @@ mod tests {
         tag_file.seek(io::SeekFrom::Start(0)).unwrap();
         assert!(!Tag::remove_from(&mut tag_file).unwrap());
     }
+
+    #[test]
+    fn test_junk_data() {
```
Comment (polyfloyd) [inline src/tag.rs:1433]: For this test, could you please:

* Rename this test to test_issue_39?
* Add a link to #39?
* Copy quiet.mp3 to a temp file and write to that? Tip: `tempfile::NamedTempFile::new()`
* Spin up ffprobe and check its output for whether the error is absent?

File: src/stream/tag.rs [hunk 2]
```diff
@@ -226,7 +226,7 @@ impl Encoder {
     ///
     /// Note that the plain tag is written, regardless of the original contents. To safely encode a
     /// tag to an MP3 file, use `Encoder::encode_to_path`.
-    pub fn encode(&self, tag: &Tag, mut writer: impl io::Write) -> crate::Result<()> {
+    pub fn encode(&self, tag: &Tag, mut writer: impl io::Write, padding: Option<usize>) -> crate::Result<()> {
```
Comment (polyfloyd) [inline src/stream/tag.rs:229]: Encoder has a complementary Builder, could you please add the padding argument to that? That way we can keep the API backwards compatible :)

File: src/stream/tag.rs [hunk 3]
```diff
@@ -273,7 +273,7 @@ impl Encoder {
 
         let mut storage = PlainStorage::new(file, location);
         let mut w = storage.writer()?;
-        self.encode(tag, &mut w)?;
+        self.encode(tag, &mut w, Some(1024))?;
```
Comment (polyfloyd) [inline src/stream/tag.rs:276]: Hardcoding this number spooks me quite a bit, we should make some effort to de-duplicate it.

The padding is also written by the storage module which makes this an interesting challenge to solve in an correct manner, hmmm...

The storage module is also able to grow and shrink the tag padding as it desires so this number will only be correct for some cases. But the padding can only be determined after the tag has been serialized which would require the storage module to either have knowledge about the ID3 format to update the header or depend on ID3 serialization to determine the padding...

**Review by polyfloyd — CHANGES_REQUESTED** [review]
Ah! So you are suggesting it's the padding that should also be counted towards the encoded tag size.

I have added some comments. There is one bit that is going to be challenging to solve, so if you are not able to solve that yourself you can leave the rest to me to figure out :)

Thank you for your effort so far! :+1:

File: src/stream/tag.rs (hunk 3 above)
Comment (Marekkon5) [reply src/stream/tag.rs:276]: Yeah I also agree that hardcoding is dirty, but before it was hardcoded to 2048 in the storage module.

**Comment by Marekkon5** [comment]
Hello, I've made changes to the test and added the padding to the encoder builder. I've also included ffmpeg in the Github Actions workflow, so the test can pass.

However I didn't solve the last problem, as I am not really familiar with the entire codebase to do it properly. Hopefully you can fix that and merge soon. Thank you.

**Comment by Marekkon5** [comment]
I've actually tested it right now on different (actual music files) and the issue still seems to be present.

**Comment by Marekkon5** [comment]
Okay, so this commit was tested on ~1000 MP3 files, all were issue free, so hopefully that's it.
However I had to change the assert in `storage.rs` on line 263, which am not completely sure what does. However without the change it crashes after trying to write multiple times to same file, and with the change I didn't have any issues, so I am assuming it's safe, but let me know. Thank you.

File: src/stream/tag.rs (hunk 3 above)
Comment (polyfloyd) [reply src/stream/tag.rs]: The 2048 is determined by a fair dice roll, but was only defined there so hardcoding a default was fine. Having 1024 here and 1023 there creates an implicit relation which is bad practice.

The original value is used as a 'preferred' padding which is used for when the new tag no longer fits within the padded area and the file needs to be rewritten. This parameter controls the size of the padding that is added. The padding size therefore is not necessarily equal to this number.

**Comment by polyfloyd** [comment]
Thanks! I try to schedule some time to push this along the coming weekend :)

**Comment by polyfloyd** [comment]
Reviewed and merged! Although I have squashed your changes and fixed to off-by-one and storage padding issues in another commit :)

Once again, thank you for your effort!

**Pull request closed without merging**